File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:

    vector<vector<int>>ans;
    vector<int>seq;
    int n;

    void solution(int i,vector<int>&arr,int tgt){
        if(i==n){
            if(tgt==0) ans.emplace_back(seq);
            return;
        }

        if(tgt<0) return;

        seq.emplace_back(arr[i]);
        solution(i,arr,tgt-arr[i]);
        seq.pop_back();
        solution(i+1,arr,tgt);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        n=candidates.size();
        solution(0,candidates,target);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {

    void solution(int i,int tgt,List<Integer>seq,int candidates[],List<List<Integer>>ans){
        if(i==candidates.length){
            if(tgt==0){
                ans.add(new ArrayList<>(seq));
            }
            return;
        }

        if(tgt<0) return;

        seq.add(candidates[i]);
        solution(i,tgt-candidates[i],seq,candidates,ans);
        seq.remove(seq.size()-1);
        solution(i+1,tgt,seq,candidates,ans);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>>ans=new ArrayList<>();
        List<Integer>seq=new ArrayList<>();
        solution(0,target,seq,candidates,ans);
        return ans;
    }
}